### PR TITLE
feat: Searching and scrolling in dropdowns

### LINKED ISSUE #204: searchable and scrollable dropdowns for long lists
## Use Case

Requested by @gouthamr321 

Some collections generated by EMT have ~140 datasets in them. Currently, the dropdown exceeds the height of the page, and the width cuts off the names of the datasets.

image

Adding a max height + scrolling within the dataset list would fix the height issue, but adding a search bar or search-like functionality would make it easier to navigate.

## Acceptance Criteria

- Increase width of dropdown so more of the dataset name is visible
- Add search bar and filtering across datasets
- Enforce max height on dropdown so there's a scrollable area

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Closes #204, "searchable and scrollable dropdowns for long lists."

Solution
========
- Adds a search bar and scrolling to `SelectionDropdowns` that exceed a certain number of items.
  - Search input uses fuzzy searching (using `fuse.js`) on the list of provided items.
  - Scrolling uses scroll shadows to indicate information above/below the fold.
  - Clicking on the dropdown button will focus the search input.

- Behavioral changes to dropdowns based on feedback from Lyndsay:
  - Selecting an item clears the search area and closes the dropdown.

## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. A setup step / beginning state
1. What to do next
1. Any other instructions
1. Expected behavior
1. Suggestions for testing

Screenshots (optional):
-----------------------
Show-n-tell images/animations here

Keyfiles (delete if not relevant):
-----------------------
1. main file/entry point
2. other important file

Thanks for contributing!

### CODE REVIEW

File: src/colorizer/utils/react_utils.ts [hunk 1]
```diff
@@ -72,6 +73,23 @@ export function excludeUndefinedValues<T extends Object>(obj: T): Partial<T> {
...
+  pointer-events: none;
+  // Fill the parent completely so we can overlay the
+  // shadow effects above the content.
+  top: 0;
+  left: 0;
+  width: 100%;
+  height: 100%;
+  transition: box-shadow 0.1s ease-in;
+`;
+
```
Comment (ShrimpCryptid) [inline src/colorizer/utils/react_utils.ts:92]: Copied from `FeatureThresholdsTab`

File: src/components/Dropdowns/AccessibleDropdown.tsx [hunk 2]
```diff
@@ -9,8 +9,14 @@ type AccessibleDropdownProps = {
...
-  /** Contents to be rendered inside the dropdown. */
-  dropdownContent: ReactElement;
+  /**
+   * Contents to be rendered inside the dropdown. Can either be a React element or a function
+   * that returns a React element.
+   *
+   * @param setOpenState: A callback that can be used to force the dropdown open or closed.
+   * Useful for closing the dropdown after a selection has been made.
+   */
+  dropdownContent: ReactElement | ((setOpenState: (open: boolean) => void) => ReactElement);
```
Comment (toloudis) [inline src/components/Dropdowns/AccessibleDropdown.tsx:19]: Something hard to parse about this...  dropdownContent can take a function that returns a react element, but the function takes as its argument another function that is a setter based on whether the dropdown is open or closed?  I'd consider factoring that out as a special type just for readability.

File: src/components/Dropdowns/SelectionDropdown.tsx [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
+import LoadingSpinner from "../LoadingSpinner";
 import AccessibleDropdown from "./AccessibleDropdown";
-import DropdownItem, { DropdownItemList } from "./DropdownItem";
+import DropdownItem from "./DropdownItem";
+import DropdownItemList from "./DropdownItemList";
+
+// TODO: Have the dropdown show a loading indicator after a selection has been made
+// but before the prop value updates. -> this is especially noticeable when slow datasets.
+// Is there a way we can do this using async promises, maybe? If the promise rejects,
+// discard the changed value?
```
Comment (toloudis) [inline src/components/Dropdowns/SelectionDropdown.tsx:17]: what prop value takes so long to update after a dropdown selection is made?

File: src/components/Dropdowns/SelectionDropdown.tsx [hunk 4]
```diff
@@ -26,16 +37,23 @@ type SelectionDropdownProps = {
...
   /** Width of the dropdown. Overrides the default sizing behavior if set. */
   width?: string | null;
-  style?: React.CSSProperties;
+  /**
+   * Whether the search bar should be enabled. If enabled, will show search bar and filter
+   * by search input when the total number of items is above `searchThresholdCount`. True by default.
+   */
+  enableSearch?: boolean;
+  /** The number of items that must be in the original list before the search bar will be shown. 10 by default.*/
+  searchThresholdCount?: number;
```
Comment (toloudis) [inline src/components/Dropdowns/SelectionDropdown.tsx:46]: Nit: it seems like these two could be just one single thing. like if enableSearchThresholdCount === -1 (default value?) then force disable search.  idk, maybe it's better with two separate

File: src/components/Dropdowns/SelectionDropdown.tsx [hunk 5]
```diff
@@ -75,29 +98,92 @@ export default function SelectionDropdown(inputProps: SelectionDropdownProps): R
     return "";
   }, [props.selected, items]);
 
+  // Set up fuse for fuzzy searching
+  const fuse = useMemo(() => {
+    return new Fuse(items, {
+      keys: ["key", "label"],
+      isCaseSensitive: false,
+      shouldSort: true,
```
Comment (toloudis) [inline src/components/Dropdowns/SelectionDropdown.tsx:106]: sorts alphabetically?  is that only on typed search results?

File: src/components/Dropdowns/SelectionDropdown.tsx (hunk 3 above)
Comment (ShrimpCryptid) [reply src/components/Dropdowns/SelectionDropdown.tsx:17]: [Visible in the preview link here when switching datasets](https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-244/?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fmicroscopy%2FClusterOutput%2FH2B_Deliverable_AnalysisPipelineOutput%2FH2B_Deliverable_InputImages_123Timelapses_Composite_Output%2Fcolorizer%2Fcollection.json&dataset=3500005807_2&feature=R0Nuclei%20DNA%20mean%20pixel%20intensity&color=matplotlib-cool&palette-key=adobe&bg-sat=100&bg-brightness=100&fg-alpha=100&outlier-color=c0c0c0&outlier-mode=1&filter-color=dddddd&filter-mode=1&tab=track_plot&scalebar=1&timestamp=1&path=1&keep-range=0&scatter-range=all)

The datasets dropdown is the only real offender for this; I think showing a loading indicator would be helpful, but we could also try and speed up the dataset load time. It's likely taking longer because of the number of features. See #20!

File: src/components/Dropdowns/SelectionDropdown.tsx (hunk 5 above)
Comment (ShrimpCryptid) [reply src/components/Dropdowns/SelectionDropdown.tsx:106]: `shouldSort` sorts by the matching score, i.e. how close the match is to the search string. `isCaseSensitive` and `shouldSort` are set to these values by default, but I originally explicitly set their values so it would be more obvious that that behavior was taking place. I can probably just remove them.

File: src/components/Dropdowns/SelectionDropdown.tsx (hunk 3 above)
Comment (toloudis) [reply src/components/Dropdowns/SelectionDropdown.tsx:17]: I see.  My first thought it that feels like something that should be outside of the selection dropdown's purview.  If making a selection starts loading a dataset, and loading a dataset is a long running process, the rest of the app probably needs to know about it. Anyway it's a good TODO!

File: src/components/Dropdowns/SelectionDropdown.tsx (hunk 5 above)
Comment (toloudis) [reply src/components/Dropdowns/SelectionDropdown.tsx:106]: Looks fine to me, just wanted to understand what it actually means :)

File: src/components/Dropdowns/AccessibleDropdown.tsx (hunk 2 above)
Comment (ShrimpCryptid) [reply src/components/Dropdowns/AccessibleDropdown.tsx:19]: Fixed!

File: src/components/Dropdowns/SelectionDropdown.tsx (hunk 4 above)
Comment (ShrimpCryptid) [reply src/components/Dropdowns/SelectionDropdown.tsx:46]: I like keeping them separate so disabling the searchbar feels a little more intuitive:

```
<SelectionDropdown enableSearch={false} />
vs.
<SelectionDropdown searchThresholdCount={-1}>
```

**Review by toloudis — APPROVED** [review]

**Review by rugeli — APPROVED** [review]
I've been learning more about hooks and prop design patterns to better follow along with the code and the conversation. I managed to answer my own basic questions upon further review. In addition, the demo is explanatory and helpful, thank you!
